go/engine/device_clone_token.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+// DeviceCloneToken creates a new token and sends it (along
+// with the previously created token) to the server in an effort
+// to identify possible cloning of devices.
+
+package engine
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+
+	"github.com/keybase/client/go/libkb"
+)
+
+type DeviceCloneTokenEngine struct {
+	libkb.Contextified
+}
+
+// NewDeviceCloneTokenEngine creates a new DeviceCloneTokenEngine.
+func NewDeviceCloneTokenEngine(g *libkb.GlobalContext) *DeviceCloneTokenEngine {
+	return &DeviceCloneTokenEngine{Contextified: libkb.NewContextified(g)}
+}
+
+// Name is the unique engine name.
+func (e *DeviceCloneTokenEngine) Name() string {
+	return "DeviceCloneTokenEngine"
+}
+
+// Prereqs returns the engine prereqs.
+func (e *DeviceCloneTokenEngine) Prereqs() Prereqs {
+	return Prereqs{Device: true}
+}
+
+// RequiredUIs returns the required UIs. There are none.
+func (e *DeviceCloneTokenEngine) RequiredUIs() []libkb.UIKind {
+	return nil
+}
+
+// SubConsumers returns the other UI consumers for this engine. There are none.
+func (e *DeviceCloneTokenEngine) SubConsumers() []libkb.UIConsumer {
+	return nil
+}
+
+func (e *DeviceCloneTokenEngine) getToken(m libkb.MetaContext) libkb.DeviceCloneToken {
+	env := m.G().GetEnv()
+	un := env.GetUsername()
+	return env.GetConfig().GetDeviceCloneToken(un)
+}
+
+func (e *DeviceCloneTokenEngine) persistToken(m libkb.MetaContext, d libkb.DeviceCloneToken) error {
+	env := m.G().GetEnv()
+	un := env.GetUsername()
+	return env.GetConfigWriter().SetDeviceCloneToken(un, d)
+}
+
+func (e *DeviceCloneTokenEngine) ClonesCount(m libkb.MetaContext) int {
+	count := e.getToken(m).Clones
+	if count < 1 {
+		count = 1
+	}
+	return count
+}
+
+// Run starts the engine.
+func (e *DeviceCloneTokenEngine) Run(m libkb.MetaContext) error {
+	d := e.getToken(m)
+	p, s := d.Prior, d.Stage
+	if p == "" {
+		//first run
+		p = libkb.DefaultCloneTokenValue
+	}
+	if s == "" {
+		buf := make([]byte, 16)
+		rand.Read(buf)
+		s = hex.EncodeToString(buf)
```

Comment:
we have helper functions for this.  see libkb/utils.go

Reply:
👍